external-crates/move/crates/move-compiler/src/typing/deprecation_warnings.rs

```diff
@@ -0,0 +1,564 @@
+    // ```text
+    // #[deprecated]
+    // module 0x42::m {
+    //   fun f() { }
+    //   fun g() { f(); } // call to f will _not_ emit a deprecated warning since this is within 0x42::m
+    // }
+    // ```
+    // Note that we will always emit a deprecated warning for any calls to function that is
+    // annotated as deprecated, regardless of whether the call is within the module or not:
+    // ```text
+    // module 0x42::m {
+    //   #[deprecated]
+    //   fun f() { }
+    //   fun g() { f(); } // call to f will emit a deprecated warning
+    // }
+    // ```
+    fn create_deprecation_warnings(self) {
+        for (DeprecationWarningId { deprecation, .. }, locs) in self
+            .deprecation_warnings
+            .into_iter()
+            .filter(|(deprecation_id, _)| {
+                // Dont' register a deprecation warning if the module is already deprecated and we are
+                // calling from within the deprecated module.
+                !(deprecation_id.deprecation.module_ident == deprecation_id.call_site.0
+                    && deprecation_id.deprecation.location == AttributePosition::Module)
+            })
+        {
+            let mut locs: Vec<_> = locs
+                .into_iter()
+                .flat_map(|(position, locs)| locs.into_iter().map(move |loc| (position, loc)))
+                .collect();
+            locs.sort_by_key(|(_, loc)| *loc);
+            locs.reverse();
+
+            let initial_diag = locs.pop().expect("ICE: locs should not be empty");
+
+            let module_msg = if deprecation.location == AttributePosition::Module {
+                ". It is deprecated since its whole module is marked deprecated"
+            } else {
+                ""
+            };
+
+            let location_string = match deprecation.location {
+                AttributePosition::Module => {
+                    format!("This {} is deprecated{}", initial_diag.0, module_msg,)
+                }
+                x => format!("This {} is deprecated", x),
+            };
+
+            let initial_message = match deprecation.deprecation_note {
+                None => location_string,
+                Some(note) => format!("{location_string}: {note}"),
+            };
+
+            let mut diag = diag!(
+                TypeSafety::DeprecatedUsage,
+                (initial_diag.1, initial_message)
+            );
+
+            for (position, loc) in locs {
+                diag.add_secondary_label((loc, format!("Deprecated {position} also used here")));
+            }
+            self.env.add_diag(diag);
+        }
+    }
+}
+
+// If `ty` is a qualified datatype name, return the module ident, datatype name, and location.
+// Otherwise, return None.
+fn qualified_datatype_name_of_type(ty: &Type) -> Option<(&ModuleIdent, &DatatypeName)> {
+    match &ty.value {
+        Type_::Apply(_, sp!(_, N::TypeName_::ModuleType(mident, name)), _) => Some((mident, name)),
+        _ => None,
+    }
+}
```

Comment:
This functionality exists as `ty.unfold_to_type_name().datatype_name()`, which will also walk through refs.